cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+        // filter
+        if constexpr (cuda::std::is_same_v<T, cudf::string_view> and timestamp_is_int96) {
+          auto const int128_key = static_cast<__int128_t>(d_scalar.value<int64_t>());
+          cudf::string_view probe_key{reinterpret_cast<char const*>(&int128_key), 12};
+          results[row_group_idx] = filter.contains(probe_key);
+        } else {
+          // Query the bloom filter and store results
+          results[row_group_idx] = filter.contains(d_scalar.value<T>());
+        }
+      });
+
+    return std::make_unique<cudf::column>(cudf::data_type{cudf::type_id::BOOL8},
+                                          static_cast<cudf::size_type>(num_row_groups),
+                                          std::move(results),
+                                          rmm::device_buffer{},
+                                          0);
+  }
+
+  // Creates device columns from bloom filter membership
+  template <typename T>
+  std::unique_ptr<cudf::column> operator()(cudf::size_type equality_col_idx,
+                                           cudf::data_type dtype,
+                                           ast::literal* const& literal,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr) const
+  {
+    // List, Struct, Dictionary types are not supported
+    if constexpr (cudf::is_compound<T>() and not std::is_same_v<T, string_view>) {
+      CUDF_FAIL("Compound types don't support equality predicate");
+    } else if constexpr (cudf::is_timestamp<T>()) {
+      if (parquet_types[equality_col_idx] == Type::INT96) {
+        // For INT96 timestamps, use cudf::string_view type and set timestamp_is_int96
+        return query_bloom_filter<cudf::string_view, true>(
+          equality_col_idx, dtype, literal, stream, mr);
+      } else {
+        return query_bloom_filter<T>(equality_col_idx, dtype, literal, stream, mr);
+      }
+    } else {
+      return query_bloom_filter<T>(equality_col_idx, dtype, literal, stream, mr);
+    }
+  }
+};
+
+/**
+ * @brief Collects lists of equality predicate literals in the AST expression, one list per input
+ * table column. This is used in row group filtering based on bloom filters.
+ */
+class equality_literals_collector : public ast::detail::expression_transformer {
+ public:
+  equality_literals_collector() = default;
+
+  equality_literals_collector(ast::expression const& expr, cudf::size_type num_input_columns)
+    : _num_input_columns{num_input_columns}
+  {
+    _equality_literals.resize(_num_input_columns);
+    expr.accept(*this);
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::literal const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::literal const& expr) override
+  {
+    _bloom_filter_expr = std::reference_wrapper<ast::expression const>(expr);
```

Comment:
I think I'm missing something here. It looks like this can handle equality conditions for multiple columns, but only one bloom filter expression? How are the other columns' bloom filters handled?

Reply:
We want to convert each `col == literal` expression to a corresponding bloom filter expression (a `NOT(NOT(table[col])`; table contains `bf.contains(literal)` values for each predicate column and all row groups) and process the rest of the expression tree as is.

Reply:
It might help me to talk this through with you. I am a little lost in the explanation.

Reply:
Yeah of course, let's sync offline.

Reply:
We chatted offline. The main question I had was why we're setting `_bloom_filter_expr` in every `visit` overload. It seems like the only thing we care about is tracking the root of the AST?